### PR TITLE
Add version prefix before JS caching

### PR DESCRIPTION
**Pull request by skjnldsv**
Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

### CODE REVIEW

File: lib/private/Template/JSCombiner.php [hunk 1]
```diff
@@ -256,4 +256,19 @@ public function resetCache() {
...
 	}
+
+	/**
+	 * Prepend hashed app version hash
+	 * @param string $jsFile
+	 * @param string $appId
+	 * @return string
+	 */
+	private function prependVersionPrefix(string $jsFile, string $appId): string {
+		$appVersion = \OC_App::getAppVersion($appId);
```
Comment (rullzer) [inline lib/private/Template/JSCombiner.php]: Please inject the appmanager instead of this static code

**Review by rullzer — DISMISSED** [review]

**Review by kyrofa — APPROVED** [review]
The problem caused by the issue being fixed here is preventing the snap from upgrading from v12 to v13. To ensure I tested the correct thing, I cherry-picked the commits from this PR onto stable13 and tested the upgrade. It works beautifully!

**Comment by rullzer** [comment]
@skjnldsv why is this needed btw? We have upgrade steps now that clear the caches.

**Comment by skjnldsv** [comment]
@rullzer apparently not enough since 13 still have the issue :/

File: lib/private/Template/JSCombiner.php (hunk 1 above)
Comment (skjnldsv) [reply lib/private/Template/JSCombiner.php]: TemplateLayout does not use DI, so I'm not sure it's possible here.

Comment (rullzer) [reply lib/private/Template/JSCombiner.php]: Sure it is. See all the things we inject already in the constructor.

Comment (skjnldsv) [reply lib/private/Template/JSCombiner.php]: Ah yes, I looked at the wrong place, the file is properly initiated in Serper.php, my bad :)

**Comment by juliushaertl** [comment]
> @skjnldsv why is this needed btw? We have upgrade steps now that clear the caches.

I also don't get it, after running an upgrade with `occ upgrade` the js app data folders should be empty and being generated new once a user visits the page again.

**Comment by skjnldsv** [comment]
@juliushaertl Yep :/
Does it trigger by web update as well?

**Review by juliusknorr — APPROVED** [review]

**Comment by juliushaertl** [comment]
I'm fine with this, since it will at least fix the issue for the snap. @rullzer Is that ok for you as well?

But it would make sense to investigate why the jobs are not working on update then. @kyrofa Maybe you could provide me with some commands, how I can get the snaps running and do an upgrade there to reproduce it? That would be great.

**Comment by kyrofa** [comment]
Ah ha! As I was typing out the directions to @juliushaertl, I figured out why this is happening: the upgrader is run under a confinement profile that doesn't allow access to the data. I didn't realize `occ` actually did cache wiping (since it never worked for me, haha), I thought it happened after the update on first visit or something. Anyway, since the upgrader can't write to the cache, it can't wipe it. It didn't spit an error because of this, it just seemed to skip that step, so I only just now caught on. I've adjusted the confinement, and things seem to be working now even without this PR.

**Comment by skjnldsv** [comment]
Failure unrelated: 
`Error response from daemon: Conflict. The container name "/0_6311326110416211053_clone_0" is already in use by container "880888e914b7e3ea7e02079b2ed162ab1b731a28896c4fbdb8c3f9859d10c828". You have to remove (or rename) that container to be able to reuse that name.`

**Comment by rullzer** [comment]
@skjnldsv as @kyrofa pointed out in https://github.com/nextcloud/server/pull/9367#issuecomment-387230249 that it was a config error in the snap. I would propose to just kill this PR.

**Pull request closed without merging**